src/core/polyhedral/codegen_llvm.cc

```diff
@@ -294,6 +303,88 @@ class CodeGen_TC : public Halide::Internal::CodeGen_X86 {
 
     value = sym_get(name);
   }
+
+ public:
+  void optimize_module() {
+    Halide::Internal::debug(3) << "Optimizing module\n";
+
+    if (Halide::Internal::debug::debug_level() >= 3) {
+#if LLVM_VERSION_MAJOR >= 5
+      module->print(llvm::dbgs(), nullptr, false, true);
+#else
+      module->dump();
+#endif
+    }
+
+    // We override PassManager::add so that we have an opportunity to
+    // blacklist problematic LLVM passes.
+    class MyFunctionPassManager : public llvm::legacy::FunctionPassManager {
+     public:
+      MyFunctionPassManager(llvm::Module* m)
+          : llvm::legacy::FunctionPassManager(m) {}
+      virtual void add(llvm::Pass* p) override {
+        Halide::Internal::debug(2)
+            << "Adding function pass: " << p->getPassName().str() << "\n";
+        llvm::legacy::FunctionPassManager::add(p);
+      }
+    };
+
+    class MyModulePassManager : public llvm::legacy::PassManager {
+     public:
+      virtual void add(llvm::Pass* p) override {
+        Halide::Internal::debug(2)
+            << "Adding module pass: " << p->getPassName().str() << "\n";
+        llvm::legacy::PassManager::add(p);
+      }
+    };
+
+    MyFunctionPassManager function_pass_manager(module.get());
+    MyModulePassManager module_pass_manager;
+
+    std::unique_ptr<llvm::TargetMachine> TM =
+        Halide::Internal::make_target_machine(*module);
+    module_pass_manager.add(llvm::createTargetTransformInfoWrapperPass(
+        TM ? TM->getTargetIRAnalysis() : llvm::TargetIRAnalysis()));
+    function_pass_manager.add(llvm::createTargetTransformInfoWrapperPass(
+        TM ? TM->getTargetIRAnalysis() : llvm::TargetIRAnalysis()));
+
+    llvm::PassManagerBuilder b;
+    b.OptLevel = 3;
```

Comment:
can you create constant variables for them in the header file and use that? we shouldn't hardcode magical numbers in the code :)  For example, in the header file:

```
static constexpr int kOptLevel = 3;
```